src/shogun/preprocessor/KernelPCA.cpp

```diff
@@ -125,62 +119,49 @@ bool CKernelPCA::init(CFeatures* features)
 	return false;
 }
 
-
 SGMatrix<float64_t> CKernelPCA::apply_to_feature_matrix(CFeatures* features)
 {
 	ASSERT(m_initialized)
-	CDenseFeatures<float64_t>* simple_features = (CDenseFeatures<float64_t>*)features;
-
-	int32_t num_vectors = simple_features->get_num_vectors();
-	int32_t i,j,k;
-	int32_t n = m_transformation_matrix.num_cols;
-
-	m_kernel->init(features,m_init_features);
+	int32_t num_vectors = features->get_num_vectors();
+	int32_t n = m_init_features->get_num_vectors();
 
-	float64_t* new_feature_matrix = SG_MALLOC(float64_t, m_target_dim*num_vectors);
+	m_kernel->init(features, m_init_features);
+	auto kernel_matrix = m_kernel->get_kernel_matrix();
 
-	for (i=0; i<num_vectors; i++)
+	auto rows_sum = linalg::rowwise_sum(kernel_matrix);
+	for (index_t i = 0; i < n; ++i)
 	{
-		for (j=0; j<m_target_dim; j++)
-			new_feature_matrix[i*m_target_dim+j] = m_bias_vector.vector[j];
+		linalg::add_col_vec(
+		    kernel_matrix, i, rows_sum, kernel_matrix, 1.0, -1.0 / n);
+	}
 
-		for (j=0; j<n; j++)
-		{
-			float64_t kij = m_kernel->kernel(i,j);
+	SGMatrix<float64_t> W(
+	    m_transformation_matrix.matrix, m_transformation_matrix.num_rows,
+	    m_target_dim, false);
 
-			for (k=0; k<m_target_dim; k++)
-				new_feature_matrix[k+i*m_target_dim] += kij*m_transformation_matrix.matrix[(n-k-1)*n+j];
-		}
+	SGMatrix<float64_t> new_feature_matrix =
+	    linalg::matrix_prod(W, kernel_matrix, true, true);
+
+	SGVector<float64_t> bias(m_bias_vector, m_target_dim, false);
+	for (index_t i = 0; i < num_vectors; ++i)
+	{
+		linalg::add_col_vec(new_feature_matrix, i, bias, new_feature_matrix);
 	}
 
 	m_kernel->cleanup();
-	simple_features->set_feature_matrix(SGMatrix<float64_t>(new_feature_matrix,m_target_dim,num_vectors));
-	return ((CDenseFeatures<float64_t>*)features)->get_feature_matrix();
+	return new_feature_matrix;
 }
 
 SGVector<float64_t> CKernelPCA::apply_to_feature_vector(SGVector<float64_t> vector)
 {
 	ASSERT(m_initialized)
-	SGVector<float64_t> result = SGVector<float64_t>(m_target_dim);
-	m_kernel->init(new CDenseFeatures<float64_t>(SGMatrix<float64_t>(vector.vector,vector.vlen,1)),
-	               m_init_features);
-
-	int32_t j,k;
-	int32_t n = m_transformation_matrix.num_cols;
 
-	for (j=0; j<m_target_dim; j++)
-		result.vector[j] = m_bias_vector.vector[j];
+	std::unique_ptr<CFeatures> features(
+	    new CDenseFeatures<float64_t>(SGMatrix<float64_t>(vector)));
 
-	for (j=0; j<n; j++)
-	{
-		float64_t kj = m_kernel->kernel(0,j);
-
-		for (k=0; k<m_target_dim; k++)
-			result.vector[k] += kj*m_transformation_matrix.matrix[(n-k-1)*n+j];
-	}
+	SGMatrix<float64_t> result = apply_to_feature_matrix(features.get());
 
-	m_kernel->cleanup();
-	return result;
+	return SGVector<float64_t>(result.matrix, m_target_dim);
```

Comment:
I'm having some trouble with this, in this way data gets deallocated, how do I pass to SGVector without copying?